google/google.go

```diff
@@ -117,73 +131,130 @@ func (g *GoogleCloud) subscribe(opts ps.HandlerOptions, h ps.MsgHandler, ready c
 		ready <- true
 
 		b := &backoff.Backoff{
-			//These are the defaults
-			Min:    200 * time.Millisecond,
+			Min:    1 * time.Second,
 			Max:    600 * time.Second,
 			Factor: 2,
 			Jitter: true,
 		}
 
-		// create a semaphore, this is because Google PubSub will spam
-		// your service if you can't process a message
-		// and will also not handle
-		sem := semaphore.NewWeighted(int64(opts.Concurrency))
-
 		// Listen to messages and call the MsgHandler
 		for {
 			if g.shutdown {
 				break
 			}
 
-			cctx, cancel := context.WithCancel(context.Background())
-			err = sub.Receive(cctx, func(ctx context.Context, m *pubsub.Message) {
-				if serr := sem.Acquire(ctx, 1); serr != nil {
-					logrus.Errorf(
-						"pubsub: Failed to acquire worker semaphore: %v",
-						serr,
-					)
-					return
-				}
-				defer sem.Release(1)
-
-				b.Reset()
-				msg := ps.Msg{
-					ID:          m.ID,
-					Metadata:    m.Attributes,
-					Data:        m.Data,
-					PublishTime: &m.PublishTime,
-					Ack: func() {
-						m.Ack()
-					},
-					Nack: func() {
-						m.Nack()
-					},
-				}
-
-				err = h(ctx, msg)
-				if err != nil {
-					return
-				}
-
-				if opts.AutoAck {
-					m.Ack()
-				}
-			})
+			req := &pb.PullRequest{
+				Subscription: fmt.Sprintf("projects/%s/subscriptions/%s", g.projectID, subName),
+				MaxMessages:  int32(opts.Concurrency),
+			}
+
+			conn, err := g.defaultConn(context.Background())
+			if err != nil {
+				break
+			}
 
+			resp, err := conn.Pull(context.Background(), req)
 			if err != nil {
 				d := b.Duration()
 				logrus.Errorf(
-					"Subscription receive to topic %s failed, reconnecting in %v. Err: %v",
+					"pull from topic %s failed, retrying in %v. Err: %v",
 					opts.Topic, d, err,
 				)
 				time.Sleep(d)
+				continue
+			}
+
+			logr.WithCtx(context.Background()).Debugf(
+				"Google Pubsub: pulled %d messages for %s",
+				len(resp.ReceivedMessages),
+				subName,
+			)
+
+			b.Reset()
+
+			var wg sync.WaitGroup
+			for _, m := range resp.ReceivedMessages {
+				go func() {
+					wg.Add(1)
```

Comment:
The waitgrup should be incremented before creating the goroutine, since by the time the loop is over none of the goroutines might have trigered, so `wg.Wait()` would not block